riscv-c-api.md

```diff
@@ -116,5 +116,41 @@ For example:
 
 ## Intrinsic Functions
 
-Do we really have none of these?  I can't figure out
-`gcc/gcc/config/riscv/riscv-builtins.c`...
+RISC-V intrinsics allow access to RISC-V-specific instructions from C or C++ that are usually not otherwise synthesizable by compilers.
+Intrinsics typically map to a single instruction that is inlined to avoid function call overhead.
+
+Intrinsics are only available if the corresponding instructions are available (as part of the effective `-march` string).
+Some intrinsics are only available if a particular header file is included.
+RISC-V header files, that enable intrinsics, require the prefix `riscv_` (e.g. `riscv_vector.h` or `riscv_crypto.h`).
+
+RISC-V intrinsics follow the following naming rule:
+
+```
+INTRINSIC ::= PREFIX '_' MNEMONIC [ '_' RET_TYPE ]
+PREFIX ::= "__builtin_riscv" | "__rv" EXTENSION | "_rv" | "_rv32" | "_rv64"
+EXTENSION ::= ISA extension letter or string.
+MNEMONIC ::= Instruction name in the ISA extensions specification. Replace '.' with '_' (if any).
+RET_TYPE ::= Return type indication.
+```
+
+The return type is typically encoded into the intrinsic name as part of the prefix or a postfix.
+In case the return type is encoded into the prefix, then the following prefixes should be used:
+* `_rv`: for intrinsics that operate on the long data type.
```

Comment:
Unfortunately, this naming scheme is very confusing, as it's easy to confuse "_rv32_xxx" (i.e. an intrinsic that operate on 32 bit data) with RV32 devices.

There are RV32 instructions that operate on 64 bit data, so `_rv64` intrinsics doesn't have to be RV64-specific. One such instruction is `add64` from the P extension. 

There are many potential instructions that would not fit into the 32/64 bit input model. One such example is `smar64` from the P extension which take one 64 and one 32 bit argument.

My *guess* is that the prefixes have been used by some compilers to distinguish between the RV32 and the RV64 version of an intrinsic, such as `_rv32_clz` and `_rv64_clz`. However, technical limitations are not a good foundation for user-visible names. (If a compiler needs to use different names internally, this can easily he handles by aliases in a header file.)

The C standard has different formulations when it comes to reserved identifiers. `_<lowercase letters>` is less protected than `__xxx` and `_<Uppercase letters>`. See paragraph 7.1.3 of the C standard. Just to be safe, I suggest that we always use double underscores.

I suggest that we drop this and use `__rv_xxx` or `__rv<extension>_xxx` for all instructions that map to a single instruction. Things like the P extension vector variants can be used `__rv_v_xxx` or `__rv_xxx_v`, where the latter is a better match for the suggested naming scheme, where the return type is included in the name.

/ Anders Lindgren, Lead RISC-V compiler developer at IAR Systems

Reply:
Yep, the prefix is the return length and is indeed independent of the architecture. I don't know about the P extension, but the ratified extensions include plenty of `_rv32` prefix intrinsics that work on RV64; "w" instruction variants are expressed this way. For example "_rv32_ror" (rotate) translates tot `rorw[i]` on RV64 and "ror[i]" on RV32. This is not because of a technical limitation, but because 32-bit rotate is a different thing from a 64-bit rotation. My understanding is that these intrinsics work in LLVM; Which compiler is interpreting the prefix as an architecture?

Reply:
Oh, I didn't mean that a compiler would interpret it as an architecture -- I'm concerned that users might find it confusing. (Using the trivial sample set "me" -- 100% of the members of the group was confused by this...)

I wasn't aware that `_rv32_ror` maps to `ror` or `rorw` for RV32 and RV64, respectively. (The IAR compiler has support for this, but I wasn't personally involved in that part.) This is taking things one step further away from the idea that there is a one-to-one mapping between intrinsics and instructions. Well, it might be good, but the question is where to draw the line? It would make sense to have a small set of well-defined helper functions, like `__builtin_clz`, and then provide trivial intrinsics that correspond to individual instructions.

The crux of the matter is that it's time-consuming to add support for intrinsics in a number of compilers. Today, there are at least three that is in active development (gcc, clang, and IAR). By keeping it as simple as possible on the compiler side, it's easy for somebody to write a layer on top of this that can use the compiler-provided low-level intrinsics functions. This is also true if we plan to provide "emulated" versions of, say, crypto intrinsics when using a device without that extension. It's more work (and probably extremely error prone) if each compiler should provide their own variant.

I'm aware that the RISC-V architecture has been around for some time. However -- if it is successful -- much, much more core will be written for it in the future that what has been done so far. Hence, I think it's a win for everybody if we could agree upon guidelines for intrinsics to be used in the future, so that the look and feel of intrinsics provided by various extensions will be the same. (And, if backward compatibility is important, it's probably trivial to provide headers that could define the old names in terms of the new intrinsics.)

By the way, were are the intrinsics for the ratified extensions formally defined? The B specification doesn't include them, as far as I can see. I've found header files for (presumably) gcc but that doesn't feel like a sound foundation to build a future echo system on.

Reply:
(.. snip ..)
>  This is also true if we plan to provide "emulated" versions of, say, crypto intrinsics when using a device without that extension. It's more work (and probably extremely error prone) if each compiler should provide their own variant.

**Important.** As a cryptographer and co-author of the cryptography extension I would recommend against such emulation of crypto intrinsics specifically; it can easily break the constant-time security guarantees explicitly stated by these extensions (especially if they include table lookups or conditional branches).  With some other extensions it may be ok.

The elimination of timing leakage is one of the main reasons for using crypto intrinsics as AES is very difficult/slow to implement in constant time (without table lookups) without such ISA extensions. Emulated code would be basically both insecure and very slow; so not very useful.

The repo at https://github.com/rvkrypto/rvkrypto-fips/ provides such emulation (in addition to inline assembler and some built-in mappings) only for development purposes; compilers should not do that. We just did this to do quantitative analysis before the instruction encodings were fixed (that was actually one of the very last steps in the ratification process). See *"Chapter 5. Data Independent Execution Latency Subset: Zkt"* of the ratified Scalar Crypto spec for further information: https://github.com/riscv/riscv-crypto/releases

> By the way, were are the intrinsics for the ratified extensions formally defined? The B specification doesn't include them, as far as I can see. I've found header files for (presumably) gcc but that doesn't feel like a sound foundation to build a future echo system on.

I'm afraid that the intrinsics have grown mostly organically, even though RISC-V folks have worked to help to upstream them into both gcc and clang (along with other architectural support). So this repo is the closest thing to an official specification but unfortunately has had little progress as there is no little centralized management over the intrinsics.

Reply:
Oh, see the other PR about the scalar crypto intrinsics specifically. There has been quite a lot of review on that one. https://github.com/riscv-non-isa/riscv-c-api-doc/pull/23

Reply:
Thanks for the clarifications.

As this is still early days in the RISC-V world, I still think that it would be a good idea to standardize the format used by intrinsics.

A quick question: Why does intrinsics, that work with classical bit operations like `ror` and `brev8`, have *signed* parameters and return type? From a user point of view, it's more natural to use unsigned types.

Reply:
> Thanks for the clarifications.
> 
> As this is still early days in the RISC-V world, I still think that it would be a good idea to standardize the format used by intrinsic.

Yep. Unfortunately, defining instrinsics is not a work item for the ISA extension task groups, and the grand harmonization would need to be directed from the "highest level" anyway. This has not occurred, and hence it has been left mainly to the compiler people. The vector extension is different of course and in crypto we knew that people needed those so we've proposed them. But intrinsics are unfortunately not going into the ISA specs themselves as things currently are.
 
> A quick question: Why does intrinsics, that work with classical bit operations like `ror` and `brev8`, have _signed_ parameters and return type? From a user point of view, it's more natural to use unsigned types.

Yes, I agree that it's a bit weird. However, sign extension is a "universal" convention with the RISC-V ISA. For example, the load instructions `lw` and `lhu` always sign-extend (there are no unsigned load instructions afaik). Hence other "W" instructions such as RORW (the 32-bit rotation on RV64) also sign-extend the highest bit. If the return type prototype for the corresponding intrinsic was unsigned, there might be more of a risk of the compiler creating additional instructions to clear the upper half upon variable assignment or type cast.

Reply:
In the P extension, the intrinsics are part of the ISA specification. However, this extension has a different background as it comes from Andes rather than from the academic world.

Types in the intrinsics are an interesting topic, especially if the compiler can't provide additional information than the signature (which is the case when they are defined as inline function that use inline assembly). In the P extension, some instructions only use the lower 16 bits of an input register. The intrinsics are defined using `int16_t`. Unfortunately, a naive compiler would add code to ensure that the registers are sign-extended from bit 16, even though it's not needed. On the other hand, if they would use `int32_t` (or `intXLEN_t`), the compiler would add code in other situations. Hence, without additional keywords or built-in knowledge it's impossible for the compiler to do the right thing.

I'm currently lobbying for using `int16_t` as the return type for intrinsics that return a sign-extended 16 bit value, like the instruction `kaddh` do.

Using `uint32_t` should be safe. On RV32, there is no problem since registers are 32 bits. On RV64, the standard calling convention says that both signed and unsigned 32 bit values should be sign-extended to 64 bits, this interpretation is (or should be) applied to intrinsics as well.

Anyway, I think that we should define the intrinsics to be as natural as possible from the point of view of the user. Hence, instructions that make sense on the bit-level (like rotate), should be defined using unsigned types. When working in C, a user shouldn't need to be concerned how bits outside of a type are represented.

To conclude, in addition to a generic intrinsic specification, I strongly believe we need official documents on the C-level intrinsics for each extension. Personally, I wouldn't mind to have them as part of the ISA specification, but I'm satisfied as long as they are formally defined somewhere.

By they way, what is the status of the Crypto extension? I would like to add it to the IAR tools some day, but I prefer to do that once it has reached a stable state.

Reply:
(..)
> To conclude, in addition to a generic intrinsic specification, I strongly believe we need official documents on the C-level intrinsics for each extension. Personally, I wouldn't mind to have them as part of the ISA specification, but I'm satisfied as long as they are formally defined somewhere.

RVI ratified a set of new extensions at the end of last year and those are currently being incorporated with the main privileged and unprivileged specifications. The intrinsics are very unlikely to go into this edition, and the issue really needs to be raised at the chairs / technical steering group level for that to happen. RISC-V has paid staff that edits the main specs.

https://wiki.riscv.org/display/HOME/Recently+Ratified+Extensions

> By they way, what is the status of the Crypto extension? I would like to add it to the IAR tools some day, but I prefer to do that once it has reached a stable state.

Scalar Crypto is well past stable as it is fully ratified (officially v1.01). Vector crypto will come later. That's the main reason why gnu & llvm toolchain support for scalar crypto already upstreamed with it and there are intrinsics in mainline clang too. 

I'm not sure when/if this will happen to the P extension as it kind of competes with the V extension, which is the main RISC-V Architects' preferred way of doing parallelism. V was finally ratified in the same batch as scalar crypto last november.